tests/python/pants_test/engine/test_rules.py

```diff
@@ -327,12 +341,15 @@ def b_singleton() -> B:
       create_scheduler(rules)
 
     # This error message could note near matches like the singleton.
-    self.assert_equal_with_printing(dedent("""
-                                      Rules with errors: 1
-                                        (D, [C], d_from_c()):
-                                          No rule was available to compute C with parameter type A
-                                      """).strip(),
-                                    str(cm.exception))
+    self.assert_equal_with_printing(
+      dedent(
+        """\
+        Rules with errors: 1
+          Rule(func=d_from_c(), product=D, params=[C]):
+            No rule was available to compute C with parameter type A
```

Comment:
How would you feel about this reading more like:
`Rule(func=d_from_c(), product=D, params=[C]) required a parameter of type C, but no rules exist to produce type C. Available types: A, B.`

Reply:
I really like the wording Daniel's proposed here, though -- is it possible to use that without running into the issue you describe above (what is the issue you're describing above? I'm not sure I understand what you're displaying).

Reply:
Daniel's reword was for only one specific issue of not being able to compute `C`. The edge case I pointed to is of two distinct params `B` and `C` not being able to be computed.

After typing that, I suppose this would work, I think?

```@rule(pants_test.engine.test_rules:80:a_from_b_and_c(B, C) -> A) required parameters of type B and C, but no rules exist to produce type B or type C. Available types: SubA.```

I'm not sure how clear that is, though, in terms of _both_ B and C being distinct issues.

I'm also slightly concerned about how this would scale to, let's say, 5 different params.

Reply:
Hm, to be honest, I'd prefer to punt on this one. I don't understand it well enough to be confident in landing the change. I'd rather first land this PR and then we can make that followup after.

Reply:
Happy to punt, but maybe this could be phrased as:
`@rule(pants_test.engine.test_rules:80:a_from_b_and_c(B, C) -> A) required parameters of type B, C, D, E, F, but no rules exist to produce those types. Available types: SubA.`?

I agree that the available types list may get a bit long, but if it stops someone from needing to produce a dot graph and inspect it, that's maybe worthwhile...

Reply:
I'm going to punt. Note that this thread is linked to in https://github.com/pantsbuild/pants/issues/7509#issuecomment-577445094.